Encapsulate check arguments in context object

Naxesss opened the pull request:
Split off from #12751, related to #12091.

- [x] depends on https://github.com/ppy/osu/pull/12758

Refactors checks such that
- `ICheck.Run` takes a new `BeatmapVerifierContext` instead of its previous arguments
- `BeatmapVerifierContext` contains those previous arguments and more:
    - The `IBeatmap` instance
    - The `IWorkingBeatmap` instance
    - The new `InterpretedDifficulty` setting added by https://github.com/ppy/osu/pull/12758

This makes it easier to add new contextual information for use in checks (such as the new `InterpretedDifficulty` setting), as we can add it to this context instead of as new arguments to every check.

---

The other alternative we considered was DI, but it seemed wrong to make the verifiers and checks visual components, as they should not be dependent on a visual flow existing in my opinion.

Naxesss committed:
Replace `IWorkingBeatmap` arg with `BeatmapVerifierContext` in checks

This simplifies passing of contextual information by enabling addition without needing to refactor lots of classes.

See next commit for example.

Naxesss committed:
Add interpreted difficulty info to `BeatmapVerifierContext`

Enables checks to make use of the difficulty level as shown in the settings UI.

Naxesss committed:
Rename "playableBeatmap" check arg to "beatmap"

The working beatmap is now in the context, so it's easier to distinguish beatmap type, hence no need for this prefix.

peppy commented:
Requires minor conflict resolution.

Naxesss committed:
Merge branch 'master' into beatmap-verifier-context

Naxesss committed:
Keep track of local bound copy

peppy commented on osu.Game/Rulesets/Edit/Checks/CheckAudioQuality.cs:
```diff
@@ -26,13 +26,13 @@ public class CheckAudioQuality : ICheck
         };
 
-        public IEnumerable<Issue> Run(IBeatmap playableBeatmap, IWorkingBeatmap workingBeatmap)
+        public IEnumerable<Issue> Run(IBeatmap beatmap, BeatmapVerifierContext context)
```
Is it feasible to also move the `IBeatmap` inside the context?

Naxesss replied:
Yes, but `IBeatmap` is used in almost every single check so it would make things quite verbose. The idea is that this encapsulation stores info that not all checks will need.

peppy replied:
by verbose, you mean accessing `context.Beatmap` instead of `beatmap`?

smoogipoo replied:
I prefer `context.Beatmap` in any case. Less potential for breakage (this is a ruleset API after all), and keeps things in one place.

Naxesss replied:
Yeah, I'd prefer doing just `beatmap` to `context.Beatmap` as it's used a lot. I doubt we'd be likely to change `IBeatmap` as it's a pretty general interface, so the potential for breakage is already quite low, unless I'm misunderstanding.

smoogipoo replied:
I can imagine a future in which we change the parameter type from `IBeatmap` to something like `EditorBeatmap`, for example. Such change wouldn't break if it was placed in the context.

Naxesss replied:
I can only imagine the reason for doing so would be giving checks additional information about the editor itself, in which case it would probably be better to create a subclass of `ICheck` e.g. `EditorCheck` which specifically deals with the `EditorBeatmap`.

smoogipoo replied:
... Or adding common editor functions, like `GetClosestSnappedTime` which was moved there for `CheckUnsnappedObjects` to be able to use it.

`EditorBeatmap` was a bad example - it should never be exposed as is because it has mutation functions.

peppy replied:
I think I'm loosely for putting the beatmap inside the context because it feels like an arbitrary choice to leave it out. The pattern here is one used in many places in .NET via configuration objects, and in my experience I've never seen the "main" part separated out.

Naxesss replied:
Sure, we'll go with that then. I considered `BeatmapVerifierContext` more as something supplementing the beatmap instance (its name probably communicated that poorly), but after trying both approaches I don't really mind too much either way.

peppy commented on osu.Game/Rulesets/Edit/BeatmapVerifierContext.cs:
```diff
@@ -0,0 +1,31 @@
+        /// <summary>
+        /// The difficulty level which the current beatmap is considered to be.
+        /// </summary>
+        public readonly Bindable<DifficultyRating> InterpretedDifficulty;
```
If consumers of this context are not intending to handle changes in this bindable (it doesn't look like they would/should) then this should just be copied out of the bindable for clarity. Also avoids the `GetBoundCopy` being passed into a new instance, which is something we generally don't do.

Naxesss committed:
Change interpreted difficulty from bindable to regular value

There's no reason for why checks would need this to be bindable. A 1-directional binding is more appropriate.

Naxesss committed:
Move `IBeatmap` arg into context

Naxesss committed:
Remove accidental whitespace

peppy commented on osu.Game/Rulesets/Edit/BeatmapVerifierContext.cs:
```diff
@@ -0,0 +1,36 @@
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Game.Beatmaps;
+
```
```suggestion
 #nullable enable
```

Best to use this in non-drawable classes to make it explicit as to which properties should be nullable.

Naxesss committed:
Add `#nullable enable`

Co-authored-by: Dean Herbert <pe@ppy.sh>

Naxesss committed:
Fix formatting of `#nullable enable`

peppy commented on osu.Game.Tests/Editing/Checks/CheckUnsnappedObjectsTest.cs:
```diff
@@ -124,32 +125,32 @@ private Mock<Slider> getSliderMock(double startTime, double endTime, int repeats
                 ControlPointInfo = cpi,
                 HitObjects = hitobjects
-            };
+            }, null);
```
This will trigger a warning with the addition of https://github.com/ppy/osu/pull/12801.

Naxesss committed:
Use `TestWorkingBeatmap` instead of null in tests

Fixes the warning that #12801 will give.

peppy committed:
Fix case of `hitObjects` variables

peppy committed:
Add local variable for regular access to `HitObjects`

peppy committed:
Tidy bindable changed code up
